src/loaders.py

```diff
@@ -1,5 +1,162 @@
 
+from .retinanet import DataEncoder
 
-class DetectionLoader(ImageSegmentationBaseDataset):
-    pass
+
+class RandomSubsetSampler(Sampler):
+    def __init__(self, data_size, subset_size):
+        super().__init__()
+        self.data_size = data_size
+        self.subset_size = subset_size
+
+    def __iter__(self):
+        return iter(torch.randperm(self.data_size)[:self.subset_size].long())
+
+    def __len__(self):
+        return self.subset_size
+
+
+class ImageDetectionDataset(Dataset):
+    def __init__(self, X, y, labels, images_dir, encoder, train_mode):
+        super().__init__()
+        self.X = X
+        if y is not None:
+            self.y = y
+            self.labels = labels
+        else:
+            self.y = None
+            self.labels = None
+        self.images_dir = images_dir
+        self.encoder = encoder
+        self.train_mode = train_mode
+
+    def __len__(self):
+        return self.X.shape[0]
+
+    def __getitem__(self, index):
+
+        Xi = self.load_from_disk(index)
+
+        if self.y is not None:
+            yi = self.load_target(index, Xi.size)
+            return Xi, yi
+        else:
+            return Xi
+
+    def load_from_disk(self, index):
+        imgId = self.X[index]
+        img_path = os.path.join(self.images_dir, imgId + '.png')
+        return self.load_from_disk(img_path)
+
+    def load_image(self, img_filepath, grayscale=False):
+        image = Image.open(img_filepath, 'r')
+        if not grayscale:
+            image = image.convert('RGB')
+        else:
+            image = image.convert('L')
+        return image
+
+    def load_target(self, index, img_shape):
+        imgId = self.X[index]
+        boxes_rows = self.y.query('ImageID == {}'.format(imgId))
+        return self.get_boxes_and_labels(boxes_rows, img_shape)
+
+    def get_boxes_and_labels(self, boxes_rows, img_shape):
+        boxes = []
+        labels = []
+        h, w = img_shape
+        for box_row in boxes_rows:
+            x_min = box_row['XMin'].values[0] * h
+            x_max = box_row['XMax'].values[0] * h
+            y_min = box_row['YMin'].values[0] * w
+            y_max = box_row['YMax'].values[0] * w
+            label = box_row['LabelName'].values[0]
+            boxes.append([x_min, y_min, x_max, y_max])
+            labels.append(label)
+        return torch.FloatTensor(boxes), torch.LongTensor(labels)
+
+    def collate_fn(self, batch):
+        """Encode targets.
+
+        Args:
+          batch: (list) of images, cls_targets, loc_targets.
+
+        Returns:
+          images, stacked cls_targets, stacked loc_targets.
+        """
+        imgs = [x[0] for x in batch]
+        boxes = [x[1] for x in batch]
+        labels = [x[2] for x in batch]
+
+        inputs = torch.stack(imgs)
+        loc_targets = []
+        cls_targets = []
+        for i in range(len(imgs)):
+            loc_target, cls_target = self.encoder.encode(boxes[i], labels[i], input_size=inputs.size()[2:])
```

Comment:
@taraspiotr I was planning on passing a transformed data frame to the loader object rather than the encoder itself.
Is that ok ?

Reply:
@jakubczakon This is not the encoder for labels in dataframe, this encoder prepares targets from dataframe for RetinaNet, e.g. for boxes it gets tensor [#boxes_on_image, 4] and returns tensor [#anchors_per_image, 4]